### DISCUSSION TITLE
Combining database functions and JS SDK queries

### DISCUSSION THREAD

**Discussion by davidfloegel** [supabase/supabase#12254] [Questions]
Hi,

I'm building an app where the user can filter results by certain criteria. So far I have "all" and "your area".

Your area is only returning post within a radius of the users current location.

So far, I've set up the `postgis` extension, I'm storing the `POINT` on each post and it's all working fine.

What I'm not sure about is how to best run the two queries so that they return the same fields etc.

For my main posts, I'm using the SDK like so (and this will also include pagination)
```
  return supabase
    .from<PostResponse>("posts")
    .select(
      `
          *,
          user (
            id,
            username,
            avatarconfig
          ),
          reactions (
            reaction,
            post,
            user (
              id,
              username
            )
          )
          location
        `
    )
    .order("created_at", { ascending: false })
    .filter("archived_at", "is", "null");
```

For the `your area` posts I've created a database function but this seems like I'll be duplicating a lot of work by making sure both queries include the correct relations, are ordered the same and the pagination works the same.

What's the best way to solve this with supabase?

One idea I had was to have the database function only return `ids` of the right posts and then use the `sdk` to get posts where those id's match but that also doesn't seem ideal.

Regards

David

**Comment by GaryAustin1** [comment]
If you have a function that can return all rows, or limited rows based on your location as a parameter, then you can just have the rpc call return a table (or setof table type) and use .select() and filters, etc. on the rpc call...

https://postgrest.org/en/stable/api.html#function-filters 

This is the REST API format, but you can see select being used with embedded relations and then filters being used in the urls.   The same works for the clients.

**Reply by GaryAustin1** [reply]
You should not create new answers in github to respond...
In response to your below:

I believe so.  If not you could also pass in .rpc('my function',{distance:5000}).  And then have your code also deal with the parameter not being there for all.

This shows examples of filters on the rpc result... https://supabase.com/docs/reference/javascript/rpc

You can also do the embedded foreign table additions in a .select() is what I get from the PostgREST docs I linked. (NOTE see below on embedded)

Edit also you probably have to use a table return and define your columns as you are not sending back the same as the table.  This would also impact the embedded fk's most likely also.   Versus `setof mytable` which then knows what all the columns are.

**Reply by davidfloegel** [reply]
> You should not create new answers in github to respond...

Oh sorry, first time using Github discussions!

Hmm, ok I will try and see what I can do, I'm not sure I quite follow but I'll try!

Thanks for the help :)

**Reply by GaryAustin1** [reply]
If you are wanting more detail or help, versus just can you do something, you might consider the Supabase Discord as a bit more interactive and has decent user participation.

As far as the function, it will all be Postgres stuff (like returning a table) so lots on the web.

**Comment by davidfloegel** [comment]
Hi @GaryAustin1 Thanks so much for the response!

So, I hope I'm understanding this correctly, but it sounds like I should be able to do this:

1. Create a function that returns all posts with all fields I need and the calculated distance added on:
```js
select id, text, st_astext(location) as location, st_distance(location, st_point(-0.127758, 51.507351)::geography) as dist_meters
  from public.posts
  order by location;
```

Then call _that_ function like:
```js
const { data, error } = await supabase
  .rpc('my_function')
  . lt('dist_meters', 5000)
```

And I'd be good to go?
